src/time_bucket.c

```diff
@@ -462,3 +462,46 @@ ts_time_bucket_ng_date(PG_FUNCTION_ARGS)
 
 	PG_RETURN_DATEADT(date);
 }
+
+TS_FUNCTION_INFO_V1(ts_time_bucket_ng_timezone);
+TSDLLEXPORT Datum
+ts_time_bucket_ng_timezone(PG_FUNCTION_ARGS)
+{
+	Timestamp result;
+	const char *result_cstr;
+	const char *timezone_cstr;
+	size_t result_len;
+	char *final_cstr;
+	Datum timestamp;
+	Datum interval = PG_GETARG_DATUM(0);
+	Datum timestamptz = PG_GETARG_DATUM(1);
+	Datum timezone = PG_GETARG_DATUM(2);
+
+	/*
+	 * Convert 'timestamptz' to TIMESTAMP at given 'timezone'.
+	 * The code is equal to 'timestamptz AT TIME ZONE timezone'.
+	 */
+	timestamp = DirectFunctionCall2(timestamptz_zone, timezone, timestamptz);
+
+	/* Then treat resulting timestamp as a regular one */
+	result =
+		DatumGetTimestamp(DirectFunctionCall2(ts_time_bucket_ng_timestamp, interval, timestamp));
+	if (TIMESTAMP_NOT_FINITE(result))
+		PG_RETURN_TIMESTAMP(result);
+
+	/* Literally concatenate the name of the timezone to the resulting TIMESTAMP */
+	result_cstr = DatumGetCString(DirectFunctionCall1(timestamp_out, TimestampGetDatum(result)));
+	timezone_cstr = TextDatumGetCString(timezone);
+	result_len = strlen(result_cstr);
+	final_cstr = palloc(result_len + strlen(timezone_cstr) + 2 /* space and trailing \0 */);
+	strcpy(final_cstr, result_cstr);
+	final_cstr[result_len] = ' ';
+	strcpy(final_cstr + result_len + 1, timezone_cstr);
```

Comment:
Why are you not using postgres timezone function here? This would make this part much cleaner.
```
bench=# \df timezone
                                         List of functions
   Schema   |   Name   |      Result data type       |          Argument data types          | Type
------------+----------+-----------------------------+---------------------------------------+------
 pg_catalog | timezone | timestamp with time zone    | interval, timestamp without time zone | func
 pg_catalog | timezone | timestamp without time zone | interval, timestamp with time zone    | func
 pg_catalog | timezone | time with time zone         | interval, time with time zone         | func
 pg_catalog | timezone | timestamp with time zone    | text, timestamp without time zone     | func
 pg_catalog | timezone | timestamp without time zone | text, timestamp with time zone        | func
 pg_catalog | timezone | time with time zone         | text, time with time zone             | func
(6 rows)
```

Reply:
Good idea, thanks!